### PR TITLE
Prevent hyphens from appearing in @racket[...] identifiers when they …

### PR DESCRIPTION
**Pull request by LeifAndersen**
…overflow.

Fixes issue #99

### CODE REVIEW

**Comment by rfindler** [comment]
This seems like a big change, and I'd worry that it might break other things.

I don't have any concrete issues, but the change seems a bit scary to me. Not just because I think that it is okay to have hyphens (and, indeed, preferable in some cases) but also because this macro is a pretty low-level one.

I think that maybe it might be good to add an optional keyword argument to @racket[] and perhaps have a default value for it that could be set on a per-document basis?

**Comment by LeifAndersen** [comment]
> This seems like a big change, and I'd worry that it might break other things.

Ya, I get that fear. I have no idea if this would break a lot of things as there really isn't a set of good scribble tests (Yet, that PR will come eventually. :) ), but I can say it does work for my documents.

> Not just because I think that it is okay to have hyphens (and, indeed, preferable in some cases) but also because this macro is a pretty low-level one.

Huh, interesting. Can you give me a case where you would want a hyphen in the middle of an identifier or keyword? (That wasn't already there anyway. I'd be okay with doing it for identifiers that already have them.)

> I think that maybe it might be good to add an optional keyword argument to @racket[] and perhaps have a default value for it that could be set on a per-document basis?

I'm not opposed to having an optional argument. Although I would be opposed to needing to set it for every @racket[] block. I'd be down for a document wide setting, but how would you do that? (A parameter is a good start, but it would need to be set in every module.)

**Comment by rfindler** [comment]
I think the reason that one might want a hyphen is the same reason that one might want a hyphen in ordinary words. I do agree with you that usually it is not what you want. But we also don't want to go into the margin and it isn't obvious which badness is worse.

As for document-wide setting, I guess one could use state. But maybe it should be delimited, so a parameter might be better.

**Comment by LeifAndersen** [comment]
> But we also don't want to go into the margin and it isn't obvious which badness is worse.

That is true. IMO the best thing to do is to underfil the line, but even that is up for debate.

> As for document-wide setting, I guess one could use state. But maybe it should be delimited, so a parameter might be better.

So I added the SHyphen macro, which is like the SColorize one defined in the same file. That way there is a hook that writers can use to set how they want identifiers to be handled.

**Comment by mflatt** [comment]
I think the default shouldn't change. I can appreciate wanting to disable hyphenation everywhere, but it's a trade-off between a kind of precision (not including a "-" that might be mistaken as part of the identifier) and other concerns (like intraline spacing), and we shouldn't change existing documents.

**Comment by LeifAndersen** [comment]
That's fair. We could add `SHyphen` as a no-op, so that style files can define it as mbox or hyphen if they want, how does that sound?

**Comment by rfindler** [comment]
I agree with Matthew.

Thank you for making it easy to customize this behavior, tho! Would it make sense to use this as an example to make the docs better so people have more customization examples to work from? I think some code that starts by saying "here's what yo uneed to add to your @title{} to get rid of these hyphens" could be generally very useful to people as they could work from that other kinds of customizations at that level. (We need more help for people at that level in Scribble.)

**Comment by LeifAndersen** [comment]
Ya sure. I also agree that it can be an example to the docs, or it could also be a tutorial in blog post form (or both.)

**Comment by mflatt** [comment]
Adding `\SHyphen` as a documented no-op sounds good.

**Comment by LeifAndersen** [comment]
Okay cool. I'll make that change, add documentation, create a tutorial for the blog.

**Comment by LeifAndersen** [comment]
Okay, I've added documentation. I'll merge this and write up a short tutorial as a blog post. Unless anyone has any last minute thoughts or issues? (cc @rfindler and @mflatt )

File: scribble-doc/scribblings/scribble/config.scrbl [hunk 1]
```diff
@@ -640,6 +640,20 @@ style:
 
  ]
 
+Additionally, the @filepath{racket.tex} Latex configuration
```
Comment (bennn) [inline scribble-doc/scribblings/scribble/config.scrbl]: LaTeX

File: scribble-doc/scribblings/scribble/config.scrbl [hunk 2]
```diff
@@ -640,6 +640,20 @@ style:
...
+includes several macros that you can redefine to adjust the
+output style of Racket code:
+
+@itemlist[
```
Comment (bennn) [inline scribble-doc/scribblings/scribble/config.scrbl]: Both these should say what they're called with (rendered arguments of `@racket` ?) so it's clearer how to override them. Reading about `SHyphen`, it seemed like a boolean parameter until I got to the example.

File: scribble-doc/scribblings/scribble/config.scrbl (hunk 1 above)
Comment (LeifAndersen) [inline scribble-doc/scribblings/scribble/config.scrbl]: I have no opinion one way or the other. But the rest of the docs say Latex. So if we do make it LaTeX, we should change that consistently.

Comment (bennn) [inline scribble-doc/scribblings/scribble/config.scrbl]: Today can be the first day of the rest of this document's life.

Comment (LeifAndersen) [inline scribble-doc/scribblings/scribble/config.scrbl]: Let me rephrase...this should be a second PR. ;)

File: scribble-doc/scribblings/scribble/config.scrbl (hunk 2 above)
Comment (LeifAndersen) [inline scribble-doc/scribblings/scribble/config.scrbl]: Thanks, fixed.

File: scribble-doc/scribblings/scribble/config.scrbl [hunk 3]
```diff
@@ -640,6 +640,25 @@ style:
...
+  second argument is the text for that color.}
+ @item{@ltxd[1]{SHyphen} --- Enables or Disables the ability
+  for identifiers and keywords in Racket code from being
+  hyphenated. Defaults to enabled (for compatibility).
+  Redefine to disable or change the hyphenation behavior. For
+  example, to cause the text to overfill rather than hyphen,
+  it can be redefined to:
+  @tt["\\renewcommand{\\SHyphen}[1]{\\mbox{#1}}"]. The first
+  argument is an identifier or keyword inside of a code
+  block.}]
```
Comment (bennn) [inline scribble-doc/scribblings/scribble/config.scrbl]: I still don't think this documentation is clear. Here's an idea:
- remove `\SHyphen`, make a `\newif\ifSHyphen` and change `\incolor` to use the if
- export & document `\SHyphentrue` and `\SHyphenfalse`
(This only makes sense if there's just one way to enable / disable hyphens)

Comment (LeifAndersen) [inline scribble-doc/scribblings/scribble/config.scrbl]: Ah, okay. No, I'd rather leave this as is because while it makes sense to enable/disable hyphens, there's lots of different strategies that may or may not be okay for your document. You might be okay with, say, text box overflows, or you might prefer underflows, or you might want an abbreviation (although I admit I'm stretching on that one.)

**Comment by LeifAndersen** [comment]
Okay, since it seems to be okay, I'm going to merge this in.
